README.txt
emdrb
    by Rafael R. Sevilla <dido@imperium.ph>
    http://emdrb.rubyforge.org

== DESCRIPTION:

This is a distributed Ruby client and server which should work as a
drop-in replacement for the standard distributed Ruby implementation
available in the Ruby standard library.

== FEATURES/PROBLEMS:

This is a simple but working DRb client/server implementation that
uses EventMachine as its basis, rather than the default implementation
that uses traditional Ruby sockets.  This should at the very least
play better with other programs that have an EventMachine event loop,
and hopefully provide somewhat better scalability.

EMDRb already supports the following features of the standard DRb:

* Clients and servers over TCP/IP
* Clients and servers over Unix domain sockets
* Standard distributed Ruby ACLs

It also has the following extensions, which derive from its basis in
EventMachine:

* Client-side asynchronous method calls
* EventMachine-aware server methods capable of asynchrony.

It is, however, not yet a complete replacement for the standard DRb:

* No SSL support.  To support this fully, EventMachine needs to have
  more comprehensive SSL/TLS support.
* RSpec tests are very basic, and need a lot more comprehensive work.
* Many standard configuration options for DRb still unsupported

These and many other problems are scheduled to be addressed in the
following releases.

== SYNOPSIS:

EMDRb basically reopens several classes, adds methods, and overrides other
methods in the basic distributed Ruby implementation to make it use
EventMachine's infrastructure instead of the traditional Ruby networking
code.  One could do the following, which is practically identical to one
of the examples for distributed Ruby:

  require 'emdrb'

  URI = "druby://localhost:8787"

  class TimeServer
    def get_current_time
      return(Time.now)
    end
  end

  $SAFE=1
  DRb.start_service(URI, TimeServer.new)
  DRb.thread.join

The corresponding client code could be made nearly identical:

  require 'emdrb'

  SERVER_URI="druby://localhost:8787"

  DRb.start_service
   
  timeserver = DRbObject.new_with_uri(SERVER_URI)
  puts timeserver.get_current_time 

Or it could be written to use of asynchronous calls:

  require 'emdrb'

  SERVER_URI="druby://localhost:8787"

  DRb.start_service
   
  timeserver = DRbObject.new_with_uri(SERVER_URI)

  EventMachine::next_tick do
    timeserver.async_call(:get_current_time).callback do |res|
      puts res
    end
  end

== REQUIREMENTS:

* Obviously, EMDRb requires EventMachine.

== INSTALL:

* Standard gem installation: 'sudo gem install' ought to do the trick.

If you want to install it manually, you can always download it at the
EMDRb project page at:

http://rubyforge.org/projects/emdrb

== LICENSE:

Copyright © 2008, 2009 Rafael R. Sevilla.  You can redistribute it
and/or modify it under the same terms as Ruby.  Please see the file
COPYING for more details.

$Id$



History.txt
== 0.4.4 / 2010-04-20
* Further work on testing emdrb for Ruby 1.9

== 0.4.3 / 2010-02-01
* Preliminary work to making emdrb ready for Ruby 1.9

== 0.4.2 / 2009-04-22
* Fix exception propagation for local deferrable methods

== 0.4.1 / 2009-04-21
* Bug in calling deferrable methods from a local object now fixed

== 0.4.0 / 2009-03-18
* Mechanism for deferrable methods using blocks cleaned up further
* Asynchronous method call API cleaned up

== 0.3.2 / 2009-03-16
* Deferrable method mechanism cleaned up

== 0.3.1 / 2009-02-04
* Used EventMachine::defer instead of spawning own threads

== 0.3.0 / 2009-01-27

* Added provisions for DRb servers to have methods which are
  aware of EventMachine.
* Made internal changes so that other transports can be more easily supported.
* Unix domain socket transport added.
* Connection closing and stop_service are made to work.
* Added support for standard DRb ACLs.

== 0.2.0 / 2009-01-23

* DRb client implementation, including asynchronous calls.
* Changes so that DRb functionality is as untouched as possible.

== 0.1.2 / 2009-01-21

* Maintenance release

== 0.1.1 / 2008-12-02

* Basic unit tests

== 0.1.0 / 2008-10-23

* First public release



spec/drbserver.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright © 2008, 2009 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU General Public License / Ruby License
#
# $Id$
#
#----------------------------------------------------------------------------
#
# Copyright © 2008, 2009 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#
# This is the DRb server that should be run by the specs, and can execute
# using either the standard DRb or EMDRb depending on what is being tested.
#

Thread.abort_on_exception = true
if ARGV[0] == "emdrb"
  $LOAD_PATH << File.join(File.dirname(__FILE__), '../lib/')
  require 'emdrb'
elsif ARGV[0] == "drb"
  require 'drb'
else
  raise "specify emdrb or drb on the command line"
end

class TestServer
  def identity(x)
    return(x)
  end

  def addtwo(x, y)
    return(x+y)
  end

  def sum(*vals)
    return(vals.inject(0) { |x,y| x + y })
  end

  def blockyield(*vals)
    vals.each do |x|
      yield x
    end
  end

  def raise_exception
    raise "This error should be expected"
  end
end

if ARGV[0] == "emdrb"
  class TestServer
    include DRb::DRbEMSafe

    deferrable_method :block_df

    ##
    # Simple example of a deferrable method structured as a state
    # machine.  This method sums values of the data array as they
    # are returned by the caller.  The similarity of this function
    # to a tail-recursive, continuation-passing style version of
    # the same should be obvious...
    def block_df(data, state={:index => 0, :retval => 0, :df => nil}, &block)
      state[:df] ||= EventMachine::DefaultDeferrable.new
      if state[:index] >= data.length
        state[:df].set_deferred_status(:succeeded, state[:retval])
        return(state[:df])
      end

      df = yield data[state[:index]]
      df.callback do |result|
        state[:retval] += result
        state[:index] += 1
        self.block_df(data, state, &block)
      end
      df.errback do |res|
        state[:df].fail(res)
      end
      return(state[:df])
    end
  end
else
  class TestServer
    # This is a fake implementation of block_df to be used by the standard
    # DRb so that the same tests will run against either.
    def block_df(vals, &block)
      return(vals.inject(0) { |x,y| x + block.call(y) })
    end

  end
end


DRb.start_service("druby://:12345", TestServer.new)
DRb.thread.join


spec/emdrb_spec.rb
# -*- coding: utf-8 -*-
#
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright © 2008, 2009 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU General Public License / Ruby License
#
# $Id$
#
#----------------------------------------------------------------------------
#
# Copyright © 2008, 2009 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#

require File.join(File.dirname(__FILE__), %w[spec_helper])
require 'thread'

class LocalDeferrable
  include DRb::DRbEMSafe

  def df_tester(val)
    df = EventMachine::DefaultDeferrable.new
    df.succeed(val+1)
    return(df)
  end

  def df_tester_exception
    raise "test exception"
  end

  deferrable_method :df_tester
end

describe EMDRb do
  before do
    DRb.start_service("druby://localhost:56789", LocalDeferrable.new)
    @obj = DRbObject.new_with_uri("druby://localhost:12345")
    @obj2 = DRbObject.new_with_uri("druby://localhost:56789")
  end

  after do
    DRb.stop_service
  end

  it "should be able to perform simple synchronous method calls" do
    @obj.identity(1).should == 1
    @obj.addtwo(1, 2).should == 3
  end

  it "should be able to perform synchronous method calls with a block" do
    val = 1
    @obj.blockyield(1,2,3,4,5,6,7) { |x| val *= x }
    val.should == 5040
  end

  it "should propagate exceptions to the client" do
    lambda do
      @obj.raise_exception
    end.should raise_error(RuntimeError) do |error|
      error.message.should == "This error should be expected"
    end
  end

  it "should be able to perform asynchronous method calls" do
    q = Queue.new
    EventMachine::next_tick do
      @obj.send_async(:identity, 1).callback do |data|
        q << data
      end
    end
    data = q.shift
    data.should == 1
  end

  it "should be able to perform asynchronous method calls with a passed block" do
    q = Queue.new
    val = 1
    EventMachine::next_tick do
      df = @obj.send_async(:blockyield, 1,2,3,4,5,6,7) { |x| val *= x; val }
      df.callback do |data|
        q << data
      end
    end
    data = q.shift
    data[0].should be_true
    val.should == 5040
  end

  it "should work with variadic methods" do
    @obj.sum(1,2,3,4,5).should == 15
  end

  it "should use deferrable methods correctly" do
    res = @obj.block_df([1,2,3,4,5]) { |x| x }
    res.should == 15
  end

  it "should propagate exceptions raised by a block in a deferrable method" do
    lambda do
      res = @obj.block_df([1,2,3,4,5]) { |x| raise "an error" }
    end.should raise_error(RuntimeError) do |error|
      error.message.should == "an error"
    end
  end

  it "should call local deferrables correctly" do
    @obj2.df_tester(1).should == 2
    lambda { @obj2.df_tester_exception }.should raise_error(RuntimeError,
                                                            "test exception")
  end
end

# EOF


test/test_emdrb.rb
# -*- coding: utf-8 -*-
#
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright © 2008, 2009 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU General Public License / Ruby License
#
# $Id$
#
#----------------------------------------------------------------------------
#
# Copyright © 2008, 2009, 2010 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#
require 'test/unit'
require 'emdrb'
require 'thread'

class LocalDeferrable
  include DRb::DRbEMSafe

  def df_tester(val)
    df = EventMachine::DefaultDeferrable.new
    df.succeed(val+1)
    return(df)
  end

  def df_tester_exception
    raise "test exception"
  end

  deferrable_method :df_tester
end

class TestEMDRb < Test::Unit::TestCase
  def self.global_setup
    DRb.start_service("druby://localhost:56789", LocalDeferrable.new)
    # Standard DRb server instance
    @@obj = DRbObject.new_with_uri("druby://localhost:12345")
    # EMDRb server instance
    @@obj2 = DRbObject.new_with_uri("druby://localhost:54321")
    @@obj3 = DRbObject.new_with_uri("druby://localhost:56789")
    @@pid1 = fork
    if @@pid1.nil?
      exec("ruby -Ilib examples/drbserver.rb drb")
    end

    @@pid2 = fork
    if @@pid2.nil?
      exec("ruby -Ilib examples/drbserver.rb emdrb")
    end
    sleep(0.5)
  end

  def self.global_teardown
    Process.kill(:TERM, @@pid1)
    Process.kill(:TERM, @@pid2)
    Process.wait
  end

  def setup
    unless defined?(@@expected_test_count)
      @@expected_test_count = (self.class.instance_methods.reject{ |method| method[0..3] != 'test'}).length
      self.class.global_setup
    end
  end

  def teardown
    if (@@expected_test_count-=1) == 0
      self.class.global_teardown
    end
  end

  def test_simple_sync
    assert_equal(1, @@obj.identity(1))
    assert_equal(3, @@obj.addtwo(1,2))
    assert_equal(1, @@obj2.identity(1))
    assert_equal(3, @@obj2.addtwo(1,2))
  end

  def test_block_sync
    val = 1
    @@obj.blockyield(1,2,3,4,5,6,7) { |x| val *= x }
    assert_equal(5040, val)

    val = 1
    @@obj2.blockyield(1,2,3,4,5,6,7) { |x| val *= x }
    assert_equal(5040, val)
  end

  def test_exceptions
    assert_raises(RuntimeError) do
      @@obj.raise_exception
    end

    assert_raises(RuntimeError) do
      @@obj2.raise_exception
    end
  end

  def test_async
#    q = Queue.new
#    EventMachine::next_tick do
#      @@obj.send_async(:identity, 1).callback do |data|
#        q << data
#      end
#    end
#    data = q.shift
#    assert_equal(1, data)

#    q = Queue.new
#    EventMachine::next_tick do
#      @@obj2.send_async(:identity, 1).callback do |data|
#        q << data
#      end
#    end
#    data = q.shift
#    assert_equal(1, data)
  end

  def test_variadic
    assert_equal(15, @@obj.sum(1,2,3,4,5))
    assert_equal(15, @@obj2.sum(1,2,3,4,5))
  end

  def test_deferrable
    assert_equal(15, @@obj.block_df([1,2,3,4,5]) { |x| x })
    assert_equal(15, @@obj2.block_df([1,2,3,4,5]) { |x| x })

    assert_raises(RuntimeError) do
      @@obj.block_df([1,2,3,4,5]) { |x| raise "an error" }
    end

    assert_raises(RuntimeError) do
      @@obj2.block_df([1,2,3,4,5]) { |x| raise "an error" }
    end
  end

  def test_local_df
    assert_equal(2, @@obj3.df_tester(1))
    assert_raises(RuntimeError) do
      @@obj3.df_tester_exception
    end
  end

end


lib/emdrb/unix.rb
#
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright © 2008, 2009 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU General Public License / Ruby License
#
# $Id: emdrb.rb 72 2009-01-28 09:53:08Z dido $
#
#----------------------------------------------------------------------------
#
# Copyright © 2008, 2009 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#
require 'drb/unix'

module DRb

  ##
  # Unix domain socket transport
  class DRbUNIXSocket < DRbTCPSocket
    public

    def initialize(uri, config={})
      @uri = (uri.nil?) ? 'drbunix:' : uri
      @filename, @option = self.class.parse_uri(@uri)
      @filename.untaint
      @port.untaint
      @config = config
      @acl = nil
    end

    def start_server(prot)
      if @conndesc
        return(@conndesc)
      end

      lock = nil
      if @filename.nil?
        tmpdir = Dir::tmpdir
        n = 0
        while true
          begin
            tmpname = sprintf('%s/druby%d.%d', tmpdir, $$, n)
            lock = tmpname + '.lock'
            unless File.exist?(tmpname) or File.exist?(lock)
              Dir.mkdir(lock)
              @filename = tmpname
              break
            end
          rescue
            raise "cannot generate tempfile `%s'" % tmpname if n >= Max_try
            #sleep(1)
          end
          n += 1
        end
      end

      @conndesc = EventMachine::start_unix_domain_server(@filename,
                                                         prot) do |conn|
        if block_given?
          yield conn
        end
      end

      if lock
        Dir.rmdir(lock)
      end
    end

    def client_connect(prot)
      EventMachine.connect_unix_domain(@filename, prot) do |c|
        if block_given?
          yield c
        end
      end
    end

    def stop_server
      super
      File.unlink(@filename)
    end
  end

  DRbTransport.add_transport(DRbUNIXSocket)
end


lib/emdrb/version.rb
#
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright (c) 2008 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU Lesser General Public License / Ruby License
#
# $Id$
#
#----------------------------------------------------------------------------
#
# Copyright (C) 2008 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#
# EMDRb version code
#
module EMDRb
  VERSION = "0.4.4"
end


examples/drbserver.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
# Author:: Rafael R. Sevilla (mailto:dido@imperium.ph)
# Copyright:: Copyright © 2008, 2009 Rafael R. Sevilla
# Homepage:: http://emdrb.rubyforge.org/
# License:: GNU General Public License / Ruby License
#
# $Id$
#
#----------------------------------------------------------------------------
#
# Copyright © 2008, 2009 Rafael Sevilla
# This file is part of EMDRb
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of either: 1) the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version; or 2) Ruby's License.
# 
# See the file COPYING for complete licensing information.
#----------------------------------------------------------------------------
#
# This is the DRb server that should be run by the specs, and can execute
# using either the standard DRb or EMDRb depending on what is being tested.
#

Thread.abort_on_exception = true
if ARGV[0] == "emdrb"
  $LOAD_PATH << File.join(File.dirname(__FILE__), '../lib/')
  require 'emdrb'
  port = 54321
elsif ARGV[0] == "drb"
  require 'drb'
  port = 12345
else
  raise "specify emdrb or drb on the command line"
end

class TestServer
  def identity(x)
    return(x)
  end

  def addtwo(x, y)
    return(x+y)
  end

  def sum(*vals)
    return(vals.inject(0) { |x,y| x + y })
  end

  def blockyield(*vals)
    vals.each do |x|
      yield x
    end
  end

  def raise_exception
    raise "This error should be expected"
  end
end

if ARGV[0] == "emdrb"
  class TestServer
    include DRb::DRbEMSafe

    deferrable_method :block_df

    ##
    # Simple example of a deferrable method structured as a state
    # machine.  This method sums values of the data array as they
    # are returned by the caller.  The similarity of this function
    # to a tail-recursive, continuation-passing style version of
    # the same should be obvious...
    def block_df(data, state={:index => 0, :retval => 0, :df => nil}, &block)
      state[:df] ||= EventMachine::DefaultDeferrable.new
      if state[:index] >= data.length
        state[:df].set_deferred_status(:succeeded, state[:retval])
        return(state[:df])
      end

      df = yield data[state[:index]]
      df.callback do |result|
        state[:retval] += result
        state[:index] += 1
        self.block_df(data, state, &block)
      end
      df.errback do |res|
        state[:df].fail(res)
      end
      return(state[:df])
    end
  end
else
  class TestServer
    # This is a fake implementation of block_df to be used by the standard
    # DRb so that the same tests will run against either.
    def block_df(vals, &block)
      return(vals.inject(0) { |x,y| x + block.call(y) })
    end

  end
end

DRb.start_service("druby://127.0.0.1:#{port}", TestServer.new)
DRb.thread.join
